graphql.md

```diff
@@ -0,0 +1,177 @@
+# GraphQL
+
+Mutations:
+
+- [Name should begin with action type](#name-should-begin-with-action-type)
+- [Use an input object type for the argument](#use-an-input-object-type-for-the-argument)
+- [Return an object in the payload](#return-an-object-in-the-payload)
+- [Errors](#errors)
+
+## Mutations
+
+Schema in the following examples
+
+```
+type User {
+  firstName: String!
+  lastName: String!
+  age: Int!
+}
+```
+
+### Name should begin with action type
+
+✅
+```gql
+
+mutation CreateUserMutation($input: UserInputType!) {
+   createUser(input: $input) {
+       user {
+           firstName
+       }
+   }
+}
+```
+
+✅
+```gql
+
+mutation DeleteUserMutation($input: UserInputType!) {
+   deleteUser(input: $input) {
+       user {
+           firstName
+       }
+   }
+}
+```
+
+### Use an input object type for the argument
+
+✅
+```gql
+
+mutation UpdateUserMutation($input: UserInputType!) {
+   updateUser(input: $input) {
+       user {
+           firstName
+       }
+   }
+}
+```
+
+where 
+
+```
+type UserInputType {
+    firstName: String!
+    lastName: String!
+    age: Int!
+}
+
+```
+
+❌
+```gql
+mutation UpdateUserMutation($firstName: String!, $lastName: String!, $age: Int!) {
+   updateUser(firstName: $firstName, lastName: $lastName, age: $age) {
+       user {
+           firstName
+       }
+   }
+}
+```
+
+Otherwise the mutations become more cumbersome to write out as the number of arguments grow
+
+Input object types are more flexible to changes. If the arguments change, less places need to be updated
+
+In the case where different mutations have the same arguments, the input object type can be reused.
+
+```gql
+mutation CreateUserMutation($input: UserInputType!) {
+   createUser(input: $input) {
+       user {
+           firstName
+       }
+   }
+}
+```
+
+### Return an object in the payload
+
+Ideally the object you've mutated
+
+✅
+```gql
+mutation UpdateUserMutation($input: UserInputType!) {
+   updateUser(input: $input) {
+       user {
+           firstName
+           lastName
+       }
+   }
+}
+```
+
+❌
+```gql
+mutation UpdateUserMutation($input: UserInputType!) {
+   updateUser(input: $input) {
+        firstName
+        lastName
+   }
+}
+```
+
+### Errors
+
+All errors should be returned as top-level errors as stated in the [graphql specification](https://spec.graphql.org/June2018/#sec-Response-Format)
```

Comment:
Might also be worth linking to the Apollo doc as an example of what we'd like to get to https://www.apollographql.com/blog/full-stack-error-handling-with-graphql-apollo-5c12da407210?

Reply:
The error format in the article is a little different to how I've done it
```json
{
  "message": "Failed to get events due to validation errors",
  "extensions": {
    "code": "BAD_USER_INPUT",
    "exception": {
      "validationErrors": {
        "zipCode": "This is not a valid zipcode"
      }
    }
  }
}
```
the validationErrors is an object with keys corresponding to the field names whereas I return a list where you need to figure out the field name from the input path.

Which do your prefer?
Also which would work better for nested errors? i.e handling errors like those from the Quote API

Reply:
I'll spend a bit of time with this tomorrow and we can come up with a definitive answer

Reply:
Ok, so having thought about this a bit, my conclusion is I want this to be as close to the spec as possible and as usable as possible by the front end. I feel this should be made possible by communicating as much of the information as possible back from the serialiser errors where they exist - it means we don't loose functionality over using a standard REST API serialiser endpoint, and the errors produced by that object are the most complete. To that end, I think it would be good to take a look at this function (`octoenergy.interfaces.apisite.graphql.utils.formatters.format_serializer_errors`), which currently converts our serialiser errors into a type we can return via GraphQL, and this function (`src/octoenergy/interfaces/packages/core/src/utils/graphql.js parseSubmissionErrors`)which converts that into something used by the front end. Whatever interface that is easiest to work with and least lossy is fine by me - I'll leave yourself and @allycallow to work on which you think that might be 👍